databroker/src/broker.rs

```diff
@@ -1561,6 +1623,290 @@ impl<'a, 'b> AuthorizedAccess<'a, 'b> {
+                        Ok(_) => {}
+                        Err(PermissionError::Denied) => {
+                            let message = format!("Permission denied for vss_path {}", vss_path);
+                            return Err((ActuationError::PermissionDenied, message));
+                        }
+                        Err(PermissionError::Expired) => {
+                            return Err((
+                                ActuationError::PermissionExpired,
+                                "Permission expired".to_string(),
+                            ))
+                        }
+                    }
+                }
+                Err(ReadError::NotFound) => {
+                    let message = format!("Could not resolve vss_path of vss_id {}", vss_id);
+                    return Err((ActuationError::NotFound, message));
+                }
+                Err(ReadError::PermissionDenied) => {
+                    let message = format!("Permission denied for vss_id {}", vss_id);
+                    return Err((ActuationError::PermissionDenied, message));
+                }
+                Err(ReadError::PermissionExpired) => {
+                    return Err((
+                        ActuationError::PermissionExpired,
+                        "Permission expired".to_string(),
+                    ))
+                }
+            }
+        }
+
+        let actuation_changes_per_vss_id = &self
+            .map_actuation_changes_by_vss_id(actuation_changes)
+            .await;
+        for ele in actuation_changes_per_vss_id {
+            let vss_id = *ele.0;
+            let actuation_changes = ele.1.clone();
+
+            let opt_actuation_subscription = actuation_subscriptions
+                .iter()
+                .find(|subscription| subscription.vss_ids.contains(&vss_id));
+            match opt_actuation_subscription {
+                Some(actuation_subscription) => {
+                    actuation_subscription
+                        .actuation_provider
+                        .actuate(actuation_changes)
+                        .await?
+                }
+                None => {
+                    let message = format!("actuation provider for vss_id {} not available", vss_id);
+                    return Err((ActuationError::ProviderNotAvailable, message));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn actuate(
+        &self,
+        vss_id: &i32,
+        data_value: &DataValue,
+    ) -> Result<(), (ActuationError, String)> {
+        let vss_id = *vss_id;
+        let result_entry = self.get_entry_by_id(vss_id).await;
+        match result_entry {
+            Ok(entry) => {
+                let vss_path = entry.metadata.path;
+                let result_can_write_actuator =
+                    self.permissions.can_write_actuator_target(&vss_path);
+                match result_can_write_actuator {
+                    Ok(_) => {}
```

Comment:
If you are here then you can directly hand the actuate request through or not?

Reply:
Would be possible. However I tried to not mix to many concerns here. Therefore this loop just cares about the permissions. 

Logically there also would be a minor difference in both approaches. If we handle the Actuation Request inside the ok it would mean, that we might send out three actuate request and on the forth we notice that an error occurs. This could leave the database in an inconsistent state (?) because some requests are executed and some aren't. Right now the Actuation Requests are only forwarded once all permissions etc are verified.